### PR TITLE
Add pipe API in and out formatters for Features

### LINKED ISSUE #1374: Add pipe API in and out formatters for Features

### PR DESCRIPTION
**Pull request by heuermh**
Work in progress, opened pull request for review.

Fixes #1374

### CODE REVIEW

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/feature/FeatureRDD.scala [hunk 1]
```diff
@@ -121,11 +113,13 @@ object FeatureRDD {
     // cache the rdd, since we're making multiple passes
     rdd.cache()
 
-    // aggregate to create the sequence dictionary
-    val sd = new SequenceDictionary(rdd.map(getSequenceRecord)
-      .distinct
-      .collect
-      .toVector)
+    // create sequence records based on largest end coordinate
+    val featuresByContigName = rdd.keyBy(_.getContigName)
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/feature/FeatureRDD.scala]: Not sure this is a good idea, but it (partially) solves the trying to partition a sequence of length 1L problem in `GenomicRDD.pipe`

**Comment by heuermh** [comment]
I'm not sure why the counts don't add up for GTF and GFF3 when they are fine for BED and narrowPeak

```
- don't lose any features when piping as GTF format *** FAILED ***
  95 did not equal 114 (FeatureRDDSuite.scala:759)

- don't lose any features when piping as GFF3 format *** FAILED ***
  195 did not equal 199 (FeatureRDDSuite.scala:772)
```

Would we be duplicating features when partitioning?

```
scala> val gtf = sc.loadGtf("src/test/resources/Homo_sapiens.GRCh37.75.trun100.gtf")
gtf2: org.bdgenomics.adam.rdd.feature.FeatureRDD =
FeatureRDD(MapPartitionsRDD[28] at flatMap at ADAMContext.scala:1186,SequenceDictionary{
1->36081})

scala> gtf.rdd.count
res2: Long = 95

scala> val gff3 = sc.loadGff3("src/test/resources/dvl1.200.gff3")
gff3: org.bdgenomics.adam.rdd.feature.FeatureRDD =
FeatureRDD(MapPartitionsRDD[21] at flatMap at ADAMContext.scala:1166,SequenceDictionary{
1->1363541})

scala> gff3.rdd.count
res1: Long = 195
```

File: adam-core/src/test/scala/org/bdgenomics/adam/rdd/feature/FeatureRDDSuite.scala [hunk 2]
```diff
@@ -732,5 +732,56 @@ class FeatureRDDSuite extends ADAMFunSuite with TypeCheckedTripleEquals {
...
 
+  sparkTest("don't lose any features when piping as BED format") {
+    val inputPath = testFile("dvl1.200.bed")
+    val frdd = sc.loadBed(inputPath)
+    val features = frdd.rdd.count
+
+    implicit val tFormatter = BEDInFormatter
+    implicit val uFormatter = new BEDOutFormatter
+
+    val pipedRdd: FeatureRDD = frdd.pipe("tee /dev/null")
```
Comment (devin-petersohn) [inline adam-core/src/test/scala/org/bdgenomics/adam/rdd/feature/FeatureRDDSuite.scala]: Why do we use `tee /dev/null` in the test?

Comment (fnothaft) [inline adam-core/src/test/scala/org/bdgenomics/adam/rdd/feature/FeatureRDDSuite.scala]: It was the easiest one-liner I could think of that would pipe standard in to standard out unmodified without creating any artifacts.

**Comment by heuermh** [comment]
Rebased to pull in #1411, two unit tests still fail as discussed above.

**Comment by fnothaft** [comment]
@heuermh I just realized that I approved this but forgot to merge this. Is this good to go from your side? If yes, what I propose is:

* Let's merge #1422 
* Let's update this PR to add docs to the Pipe API docs noting that these new formatters exist
* Then, let's merge this PR.

How's that sound on your end?

**Comment by heuermh** [comment]
Sounds good.  I'll push a doc commit and squash after #1422 is merged.

**Comment by fnothaft** [comment]
Merged! Thanks @heuermh!

**Pull request merged**